### ISSUE TITLE
Migration step 3 fails

### ISSUE THREAD

**Issue by briri** [DMPRoadmap/roadmap#108]
I received an error when running 'db:migrate' for the 3rd step (final_schema). It complained that devise did not exist.

related to #107

**Comment by briri** [comment]
got around it with:
```shell
cp config/initializers/devise.rb.example config/initializers/devise.rb
```

I then get the following error after it runs through a bunch of the migrations:
```shell
StandardError: An error has occurred, all later migrations canceled:

uninitialized constant Org::OrganisationType
/Users/briley/Documents/workspace/roadmap/db/migrate/20161205095624_replacing_organisation_types_with_bitflags.rb:18:in `change'
NameError: uninitialized constant Org::OrganisationType
/Users/briley/Documents/workspace/roadmap/db/migrate/20161205095624_replacing_organisation_types_with_bitflags.rb:18:in `change'
Tasks: TOP => db:migrate
(See full trace by running task with --trace)
```

**Comment by briri** [comment]
Issues still exist on my second test machine.

I think our best course of action will be to create branches from those tagged commits and make corrections in the branches. Then update the documentation to have the users checkout the branch instead of the tag.

This should work since the tweaks we need to get the migrations to run are no longer relevant in the current development/master branches.

My suggestion is to do the following for each step:
```shell
git checkout aa9aa0051e6711e42b3d7fb73d8bd78cf3db1d6a
git checkout -b migration-from-dmponline-step-1
# Make any necessary changes
git push origin migration-from-dmponline-step-1
```
We would then just need to update the documentation afterward and make sure that we preserve those branches (you can 'protect' branches in the Github repo settings)

I'm happy to do this tomorrow if you guys agree that its a good approach

**Comment by vyruss** [comment]
Every scenario I went through involving history rewriting & rebasing results in problems with our repository, in the sense that these commits have already been pushed to `development` and some branches (such as @jollopre's) have been based on `development`.

Therefore I think that I agree in that the best course of action is to fix the two branches by amending whatever code exists at the .

That will be easy because we've kept them around as:
- https://github.com/DigitalCurationCentre/roadmap/tree/removing_unused_tables
- https://github.com/DigitalCurationCentre/roadmap/tree/final_schema

Upside is, if we push these branches to Roadmap, we can also push the exact same tag names, they will just be pointing to a different commit from a different branch.

**Comment by vyruss** [comment]
- Pushed [tag v0.1.9_old_schema](https://github.com/DMPRoadmap/roadmap/tree/v0.1.9_old_schema) to `DMPRoadmap/roadmap` 
- Pushed new fixed branches in `DMPRoadmap/roadmap`:
  - https://github.com/DMPRoadmap/roadmap/tree/dmponline4_upgrade_step1
    - Corresponds to [tag v0.2.0_transitional_schema](https://github.com/DMPRoadmap/roadmap/tree/v0.2.0_transitional_schema)
  - https://github.com/DMPRoadmap/roadmap/tree/dmponline4_upgrade_step2
    - Corresponds to [tag v0.2.1_new_schema](https://github.com/DMPRoadmap/roadmap/tree/v0.2.1_new_schema)
- Updated [instructions in wiki](https://github.com/DMPRoadmap/roadmap/wiki/Upgrading-from-DMPOnline_v4)

**Comment by briri** [comment]
excellent thanks @vyruss

**Comment by briri** [comment]
The first branch looks like it will work. The second doesn't address the 'uninitialized constant Org::OrganisationType' issue I mentioned above.

I'll run through though today now that we have a branch we can make the necessary adjusts and then just repoint the tag to the new commit. 

Also, should we maybe add a final step to the instructions that puts them back on the master (development currently) branch and runs db:migrate again?

**Comment by briri** [comment]
Added the OrganisationType and UsersPerm models in to fix uninitialized constant errors. Now encountering a:
```shell
StandardError: An error has occurred, all later migrations canceled:

Unknown primary key for table users_perms in model UsersPerm.
```

**Comment by briri** [comment]
Further issues with this one.
```shell
Mysql2::Error: Cannot add or update a child row: a foreign key constraint fails (`roadmap`.`#sql-6dd8_2fce`, CONSTRAINT `fk_rails_7d708f6f1e` FOREIGN KEY (`theme_id`) REFERENCES `themes` (`id`)): ALTER TABLE `themes_in_guidance` ADD CONSTRAINT `fk_rails_7d708f6f1e`
FOREIGN KEY (`theme_id`)
  REFERENCES `themes` (`id`)
/apps/dmp/apps/roadmap/roadmap/db/migrate/20161206122926_add_foreign_keys.rb:71:in `change'
```
I am using the original db/data.yml to create the starting point for the DB. I believe this was generated around the September timeframe

Logging into the DB I can see:
```shell
mysql> select * from themes_in_guidance where theme_id not in (select id from themes);
+----------+-------------+
| theme_id | guidance_id |
+----------+-------------+
|        1 |           4 |
|        2 |           2 |
|        3 |           3 |
|        4 |           1 |
+----------+-------------+
4 rows in set (0.01 sec)

mysql> select * from themes_in_guidance where guidance_id not in (select id from guidances);
+----------+-------------+
| theme_id | guidance_id |
+----------+-------------+
|        3 |           3 |
|        4 |           1 |
+----------+-------------+
2 rows in set (0.00 sec)
```

**Comment by briri** [comment]
Added step to remove orphaned records prior to creating fkeys on themes_in_guidance table: https://github.com/CDLUC3/roadmap/commit/e2e700becbbdb93ca9955bb292d21ea5d7dcaee3

Added stubs of the OrganisationType and UsersPerm models back so that we no longer receive 'uninitialized constant' errors
https://github.com/CDLUC3/roadmap/commit/60ce0eef529b390022dd5a71fdf20b2433623917

Replaced UsersPerm scrubbing section with a direct deletion of orphaned records instead: https://github.com/CDLUC3/roadmap/commit/3ef06296816e9dfa296709b6a270edc124b1d764

**Comment by briri** [comment]
Did some review of the DB after the migration and am seeing an issue:

- None of the users is associated with an Org
```shell
mysql> select count(id), org_id from users group by org_id;
+-----------+--------+
| count(id) | org_id |
+-----------+--------+
|      3114 |   NULL |
|         3 |      1 |
+-----------+--------+
2 rows in set (0.00 sec)
```

**Comment by vyruss** [comment]
@xsrust can you help out with this? I did not encounter any of the above problems when migrating the DMPonline DB

**Comment by briri** [comment]
ok based on @xsrust suggestion I took a look at the 20161021 migration file and see the rails environment check. I am not running as the test environment so it shouldn't be a factor for any of us.

The data.yml file I am using to seed the DB is from Oct. 26th and already has all of the migrations from June/July when we started working on this project (api tokens, etc.).

Jimmy sent me a dump of the data to use so I started with that and ran the following steps (after copying the data.yaml and schema.rb into my db/):
- `rake db:drop`
- `rake db:create`
- `rake db:schema:load`
- `rake db:data:load`
- `git stash`
- `git checkout v0.1.9_old_schema`
- Comment out the Postgres gem in my Gemfile
- Manually add config.secret_key to config/initializers/devise.rb
- rake db:migrate

It fails here after a while on the migration step mentioned above with:
```shell
D, [2017-03-07T15:52:07.341987 #28070] DEBUG -- :   User Exists (0.5ms)  SELECT  1 AS one FROM `users` WHERE (`users`.`email` = BINARY 'rniemi@jyu.i' AND `users`.`id` != 3151) LIMIT 1
D, [2017-03-07T15:52:07.343169 #28070] DEBUG -- :    (0.5ms)  ROLLBACK
rake aborted!

StandardError: An error has occurred, all later migrations canceled:

Validation failed: Email is not a valid email address
/apps/dmp/apps/roadmap/roadmap/db/migrate/20161021100420_single_organisation_for_users.rb:14:in `block in up'
```

That looks like it is indeed a bad email address. The question then is should we comment out our new email validator or should we get rid of the invalid data? If opt for commenting it out during the migration and then reintroduce it afterward, the user could have trouble logging in since the devise gem records things like last_sign_in_at

**Comment by vyruss** [comment]
Whoops! Totally forgot to mention we disabled the email validator for the existing database as we knew we had bad email addresses in that data, sorry about that

**Comment by briri** [comment]
so what is your plan moving forward with those bad emails?

**Comment by vyruss** [comment]
It won't be possible for users to enter bad emails from now on because of the validator. 
- On the one hand, as bad emails only exist in old databases, it might not be worth the effort writing code to fix that data when migrating. 
- On the other hand, if we're expecting lots of people to have the same problem and that's going to keep them from upgrading to Roadmap, it may be worth looking into

**Comment by briri** [comment]
Once the validator is in place it will be problematic. Anytime `user.save` is called one of those records it will fail due to the email validation. My concern is that the Devise gem works behind the scenes and manages some of the fields on the users table (e.g. last_sign_in_at, etc.). It is likely calling user.save in those cases.

**Comment by mxsstrong** [comment]
Is the email a required field?  If not, I would say delete the email fields that are invalid and re-run.  Since the emails aren’t valid, NULL is better than invalid.
If the email is required, change to a valid test email unknwonemail@something.com<mailto:unknwonemail@something.com> so they pass the validation.

**Comment by briri** [comment]
Started refactoring the migration processes. 
- upgrade process now begins at the Rails 4.2 merge point
- pulling data manipulation steps out of schema migrations and placing them into rake tasks
- adding checks for table_exists, column_exists, etc. 
- adding logging so that users have a full report of what transpired
- forcing old schema migration steps (that have been reworked) into the DB's schema_migrations table so that they are ignored

Finished up through the migration of Template and Plan structures. Have yet to remove the temporary tables and finish up the other minor tasks.

I'm working in 2 new branches:
Step1: https://github.com/CDLUC3/roadmap/commit/039bdab6f021888e1ea6abe3fb926fe0fc56bc06
Step2: https://github.com/CDLUC3/roadmap/commit/c566466e0786eec8b4a26be66c02fc6c4f6f164b

**Comment by vyruss** [comment]
After discussing with @briri and @xsrust, I've pushed these migration branches to Roadmap:

Step 1:  https://github.com/DMPRoadmap/roadmap/tree/dmponline4_upgrade_step1
Step 2: https://github.com/DMPRoadmap/roadmap/tree/dmponline4_upgrade_step2
Step 3: https://github.com/DMPRoadmap/roadmap/tree/dmponline4_upgrade_step3

TODO: 
1. Add `migrate:dataintegrity` rake task as [_step1](https://github.com/DMPRoadmap/roadmap/tree/dmponline4_upgrade_step1)'s final commit (@briri)
2. Tag these branches' last commits with v0.1.9, v.0.2.0, v.0.2.1 tags (@vyruss)
3. Rewrite migration HOWTO to better reflect new workflow (@vyruss)

**Comment by briri** [comment]
Added migrate:data_integrity check which deletes any orphaned records in the following tables:
answers_options, dmptemplates_guidance_groups, guidance_in_group, plan_sections, project_groups, project_guidance, questions_themes, themes_in_guidance, users_roles

It also removes invalid email addresses and replaces them with a unique replacement email. It logs those replacements in log/migration.log and lets the user know that it has written out to that log file.
```shell
W, [2017-03-10T12:24:53.132664 #24293]  WARN -- : Replacing invalid email address for name: rniemi@jyu.i, id: 3151, email: rniemi@jyu.i with a3edae15-91e5-4a7d-9ad1-98762e62d28b@replacement-email.org
```
These are the steps I needed to take to run step1:
- `git checkout dmponline4_upgrade_step1`
- `vi Gemfile` to comment out `pg` gem
- `rake db:migrate`
- `rake migrate:seed`
- `rake migrate:permissions`
- `rake migrate:dataintegrity`

Changes have been checked in at https://github.com/DMPRoadmap/roadmap/tree/dmponline4_upgrade_step1

**Comment by briri** [comment]
ok @xsrust as agreed, I did not create new tags. I updated the instructions and added some notes that I thought were important for pre/post upgrade. https://github.com/DMPRoadmap/roadmap/wiki/Upgrading-from-DMPonline_v4

**Comment by xsrust** [comment]
I followed the instructions on the wiki and it worked exactly as expected with no errors.
I think branches is the best approach as we can create modifications to the procedure if anything else turns up that we need to change.

**Comment by briri** [comment]
closing since there is nothing to PR. The branches are already on DMPRoadmap/roadmap development branch
